lib/grape/validations/validators/coerce.rb

```diff
@@ -63,6 +64,11 @@ def coerce_value(val)
       def type
         @option[:type].is_a?(Hash) ? @option[:type][:value] : @option[:type]
       end
+
+      def requires_coercion?(value)
+        # JSON types do not require coercion if value is valid
+        !valid_type?(value) || converter.coercer.respond_to?(:method) && !converter.is_a?(Grape::Validations::Types::Json)
```

Comment:
Can you explain `respond_to?(:method)` here?

Then, these seen unrelated, so to make it readable I would break it up.

```ruby
return true if valid_type?(value)
return true if ...
false
```

Reply:
`.method` is defined if a custom `coerce_with` had been specified by the user, or the destination `type` has a defined coercion that everything needs to go through despite validity.  I added a test that shows why this is needed, if that's the behavior we want (which is the current behavior).

The conditionals are all directly related because of the following statements:
```
If the value is not valid, we require coercion.

Otherwise (if the value is valid), we require coercion it if a `method`  has been defined and it is not JSON.
```
which is written explicitly as:
```
def requires_coercion?(value)
  if !valid_type?(value)
    return true
  else
    if converter.coercer.respond_to?(:method) && !converter.is_a?(Grape::Validations::Types::Json)
      return true
    else
      return false
    end
  end
end
```
which could go to:
```
def requires_coercion?(value)
  return true if converter.coercer.respond_to?(:method) && !converter.is_a?(Grape::Validations::Types::Json)
  return true if !valid_type?(value)
  false
end
```
which condenses to the logic given.  Is the middle one still what is preferred?

Reply:
I don't really care much, this is fine as is ;)